Implements #2635: Draws a transparent rectangle in the loop area (plus CSS exposure)

michaelgregorius opened the pull request:
The color for the rectangle can be defined in the style CSS for the
active and inactive case. Some other properties of the TimeLineWidget
are now exposed through the CSS as well:
* The color of the lines that are drawn at each tact
* The color of the tact numbers
* The font size
* The minimum height of the widget

It is therefore possible to change the height of the time line using the
CSS.

The tact numbers are drawn conditionally like the tact lines. The
numbers are drawn with a constant distance to the tact line. This gives
a more consistent picture at different zoom levels and also fixes the
broken look at very small zoom sizes like for example 12.5%.

Last but not least the order in which the elements of the time line are
drawn has been adjusted. The loop markers are now drawn after the lines,
numbers and the loop rectangle so that they overlay these elements.

Inactive loop:
![2635-loop-inactive](https://cloud.githubusercontent.com/assets/9293269/13651417/8ebc6fbc-e647-11e5-90d4-e0693ffcbc27.png)

Active loop:
![2635-loop-active](https://cloud.githubusercontent.com/assets/9293269/13651426/9b4b9604-e647-11e5-950b-5d7a37ad1ee6.png)

Small zoom:
![2635-small-zoom](https://cloud.githubusercontent.com/assets/9293269/13651433/a26b6b58-e647-11e5-9206-5eedcb243d32.png)

michaelgregorius commented:
Fixes #2635.

Umcaruje commented:
Could you also make the background color a `QBrush` qproperty?

Umcaruje commented:
This also fixes #2259

Umcaruje commented:
Actually you can replace [this](https://github.com/michaelgregorius/lmms/blob/2635-loop-area-rectangle/src/gui/TimeLineWidget.cpp#L230-L235): 
```c++
	QColor bg_color = QApplication::palette().color( QPalette::Active,
							QPalette::Background );
	QLinearGradient g( 0, 0, 0, height() );
	g.setColorAt( 0, bg_color.lighter( 150 ) );
	g.setColorAt( 1, bg_color.darker( 150 ) );
	p.fillRect( 0, 0, width(), height(), g );
```
with this:
```c++
	p.fillRect( 0, 0, width(), height(), p.background() );
```
and define the `background-color` in css with a QLinearGradient.

Umcaruje commented on src/gui/TimeLineWidget.cpp:
```diff
@@ -232,44 +235,49 @@ void TimeLineWidget::paintEvent( QPaintEvent * )
 	}
 
+	// Draw the loop rectangle
+	int loopRectMargin = 2;
```
shouldn't `loopRectMargin` and `loopRectHeight` be `const` too? Also, I think it would be better if the margin was 1 or 0 pixels.

Umcaruje commented on src/gui/TimeLineWidget.cpp:
```diff
@@ -232,44 +235,49 @@ void TimeLineWidget::paintEvent( QPaintEvent * )
+	int loopRectHeight = this->height() - 2 * loopRectMargin;
+	int const loopStart = markerX( loopBegin() ) + 8 + 1;
+	int const loopEndR = markerX( loopEnd() ) + 7;
+	int const loopRectWidth = loopEndR - loopStart;
```
The rectrangle width is off by one: 
![screenshot from 2016-03-09 23 05 06](https://cloud.githubusercontent.com/assets/6282045/13652268/89761fea-e64b-11e5-91d8-050514b96689.png)

grejppi commented:
@michaelgregorius **Tact** is not the correct word in English. Unfortunately it's used internally in the code, but in user facing things such as themes they should be called **bars**.

Umcaruje commented:
The code looks fine, and I made some observations from the screenshots. I have yet to compile it and test it out.

I actually wanted to start working on this issue today, but you ninja'd me :wink:  Next time, you should [assign](https://help.github.com/articles/assigning-issues-and-pull-requests-to-other-github-users/) yourself, so we don't start working on the same thing.

michaelgregorius commented:
@Umcaruje @grejppi Thanks for the code review! I have implemented your remarks with the last two commits.

IvanMaldonado commented:
Tested, it works for me :+1: 

OS: Manjaro 15.12 Capella
Kernel: x86_64 Linux 4.1.18-2-MANJARO

![captura de pantalla_2016-03-09_19-24-03](https://cloud.githubusercontent.com/assets/12549082/13656461/b679dce4-e62c-11e5-9496-8e53d63ae0f2.png)

IvanMaldonado commented:
I've noticed we have lost some vertical space here:

![captura de pantalla_2016-03-09_19-46-04](https://cloud.githubusercontent.com/assets/12549082/13656777/9792d13e-e62f-11e5-9661-3c88bb5bbd8e.png)

Umcaruje commented:
Yup, that looks abnormally large...

michaelgregorius commented:
Fixed by specifying the max-height of the widget as well. I have also learned that you can specify the font size only in `pt` or `px` but not in `em` (which makes sense in hindsight).

The height of the time line widget is still specified in `em` so that is scales proportional to the font size that is set.

IvanMaldonado commented:
@michaelgregorius Thank you for fixing it, yet I don't understand why this is like that now.

![captura de pantalla_2016-03-10_16-38-40](https://cloud.githubusercontent.com/assets/12549082/13687030/9287557a-e6de-11e5-8ff6-28d893cdeb1f.png)

michaelgregorius commented:
@IvanMaldonado What exact do you mean? I only see a themed version of the Piano Roll editor.

IvanMaldonado commented:
@michaelgregorius Makes more sense now?

![captura de pantalla_2016-03-10_16-47-08](https://cloud.githubusercontent.com/assets/12549082/13687231/be278e06-e6df-11e5-9e8e-8a1383e002a6.png)

michaelgregorius commented:
@IvanMaldonado Ah, ok. Yes, it was changed because the old Piano Roll window was too wide on smaller resolutions. Please check #2287 for more details.

IvanMaldonado commented:
@michaelgregorius I see... I don't know how I never noticed it but now, I'm sorry for referencing that here.

Thank you for working on this one! :+1:

Umcaruje commented on data/themes/default/style.css:
```diff
@@ -532,6 +532,22 @@ TimeLine {
+	   If you want a fixed size set min and max to the same value. */
+	min-height: 1.5em;
+	max-height: 1.5em;
+
```
You want to add this here: 
```css
	background-color: qlineargradient( x1: 0, y1: 0, x2: 0, y2: 1,
 						stop: 0 #8796a7, stop: 1.0 #3e454e );
```

to reinstate the old look. 

Also, shouldn't
```css
TimeLine {
	font-size: 8px;
}
```

be removed?

michaelgregorius replied:
I have reinstated the old look. I have to admit that I like the flat look better but that's something for theme designers I guess. :)

I have also removed the `TimeLine` entry from the `style.css`.

Umcaruje commented:
Ok, compiled and tested, and the numbers look a bit blurred on my non-high DPI monitor:
![screenshot from 2016-03-11 16 11 55](https://cloud.githubusercontent.com/assets/6282045/13706120/fd79d07c-e7a3-11e5-9705-7f45a80101dd.png)

Could you force hinting on them? Here's how I did it for TCO's:
https://github.com/LMMS/lmms/blob/08847cc7d304809a049cc9b0b70bc2fe393b4b1e/src/tracks/Pattern.cpp#L1177-L1180

michaelgregorius commented:
I have added hinting in 47e7b7d.

IvanMaldonado commented:
Can this be merged?

Umcaruje commented:
>Can this be merged?

Don't rush it please, I have a couple of remarks left, just haven't had the time to write them.

Umcaruje commented on src/gui/TimeLineWidget.cpp:
```diff
@@ -224,52 +227,59 @@ void TimeLineWidget::paintEvent( QPaintEvent * )
+	int const loopEndR = markerX( loopEnd() ) + 9;
+	int const loopRectWidth = loopEndR - loopStart;
+	p.setPen(Qt::NoPen);
+	p.setBrush(loopPointsEnabled() ? getActiveLoopColor() : getInactiveLoopColor());
```
You don't have space between parentheses on these 3 lines, while the rest of the file looks differently. It'd be nice if you changed it for consistency.

Umcaruje commented:
@michaelgregorius this all looks good, but the numbers appear off-center on my machine. 
![screenshot from 2016-04-30 17 19 40](https://cloud.githubusercontent.com/assets/6282045/14936895/15500060-0ef8-11e6-9102-cfe1b80cdf72.png)

The numbers also still look weird to me, but that's the issue of what my default font is. Ideally, we should probably choose a font to force on the app, so its consistent across all platform, but that's a completely different issue and out of scope of this PR.

Otherwise, this looks and works fine, so when you adress these 2 minor issues, I'll merge :+1:

grejppi commented:
Isn't the contrast a bit too low? When quickly looking at that screenshot @Umcaruje posted I can barely see a difference.

Umcaruje commented:
@grejppi that was a screenshot with the loop points disabled. Here's how it looks with them enabled:
![screenshot from 2016-04-30 20 09 26](https://cloud.githubusercontent.com/assets/6282045/14937679/84c7a63e-0f0f-11e6-876d-1dddce59ad06.png)

michaelgregorius commented:
@Umcaruje Done!

Here's how it looks on my machine now:
![2635-white-rectangle](https://cloud.githubusercontent.com/assets/9293269/14990493/1a302814-115d-11e6-9b52-503b13366053.png)

I think it's hard to get it perfect because you have to use font metrics and they might give different numbers for different fonts. For example the numbers of a font might not make use of the full ascent.

grejppi commented:
@michaelgregorius @Umcaruje TBH the active colour is too dim. That would still work well as the inactive colour though.

Umcaruje commented:
@grejppi but if we push the active color transparency up, the text could become hard to read. maybe a colored (green perhaps) rectangle would do good for the contrast issues?

grejppi commented:
Can't the text be drawn over the rectangle?

michaelgregorius commented:
Here is how it looks with a green loop rectangle thats drawn behind the lines and numbers.
Active:
![2635-back-green-active](https://cloud.githubusercontent.com/assets/9293269/15026596/63730f72-123f-11e6-8ad1-563b8805acf7.png)
Inactive:
![2635-back-green-inactive](https://cloud.githubusercontent.com/assets/9293269/15026601/68f18ab4-123f-11e6-8581-c4cf0ae6c431.png)
Here's the relevant CSS that was used in the screenshots:
```CSS
	qproperty-inactiveLoopColor: rgba( 0, 128, 0, 64 );
	qproperty-activeLoopColor: rgba( 0, 128, 0, 192 );
```
The intention of drawing the rectangle in front of everything was to convey more of a "coverage" feeling but I think drawing it behind everything should also serve it's purpose.

tresf commented:
The visual effect of the green is very obvious as to it's function, much more so than the gray hue. :+1:

IvanMaldonado commented:
Will the borders of the colored rectangle be editable?

grejppi commented:
@michaelgregorius how about something like...

![active_loop](https://cloud.githubusercontent.com/assets/772336/15032483/1b498980-126a-11e6-975a-9701b5c90fe4.png)

The light inner border would be consistent with clips in the song editor. Drawing the loop area as a rectangle also would make the old pixmaps redundant.

michaelgregorius commented:
@grejppi I have implemented your proposal with the latest commit. Here's how it looks now:
![2635-new-loop-active](https://cloud.githubusercontent.com/assets/9293269/15099809/605801f2-1560-11e6-93e5-53fdb2b4ace5.png)
![2635-new-loop-inactive](https://cloud.githubusercontent.com/assets/9293269/15099810/6529d142-1560-11e6-8fc5-9109b2339f84.png)
We might consider changing the Z order of some the elements, e.g. draw the lines at the very bottom.

I also had to rebase to make everything compile with GCC 6.1.

grejppi commented:
@michaelgregorius Amazing work!

As for the ordering, maybe the filled area should be drawn first, then text and bar lines, and then lastly the outlines?

michaelgregorius commented:
@grejppi Done!

![2635-z-order-changed](https://cloud.githubusercontent.com/assets/9293269/15100291/fc5f4094-156d-11e6-8497-4abc61817d23.png)
![2635-z-order-inactive](https://cloud.githubusercontent.com/assets/9293269/15100292/002c7ec6-156e-11e6-834c-0950345f9efe.png)

Umcaruje commented on src/gui/TimeLineWidget.cpp:
```diff
@@ -224,55 +215,67 @@ void TimeLineWidget::paintEvent( QPaintEvent * )
+	// Draw the inner border outline (no fill)
+	QRect innerRectangle = outerRectangle.adjusted( 1, 1, -1, -1 );
+	p.setPen( loopPointsActive ? getActiveLoopInnerColor() : getInactiveLoopInnerColor() );
+	p.setBrush(Qt::NoBrush);
```
I'd add spaces in the parentheses here, so its consistent with the rest of the file.

michaelgregorius replied:
Done with commit 83fffcc49c398a8e259609151de8c09e0175cae1.

Umcaruje commented on src/gui/TimeLineWidget.cpp:
```diff
@@ -274,17 +273,6 @@ void TimeLineWidget::paintEvent( QPaintEvent * )
-	p.drawPixmap( markerX( loopEnd() ) + 2, loopMarkerHeight, *s_loopPointEndPixmap );
-
-	p.setOpacity( 0.6 );
-	p.drawPixmap( m_posMarkerX, height() - s_posMarkerPixmap->height(),
```
@michaelgregorius you accidentaly removed the playhead marker pixmap here:
![screenshot from 2016-05-08 23 08 45](https://cloud.githubusercontent.com/assets/6282045/15100439/dfcc7966-1571-11e6-8d2d-9bd0f89e1a1f.png)

michaelgregorius replied:
@Umcaruje Thanks! Good catch! Brought back via commit 83fffcc49c398a8e259609151de8c09e0175cae1.

Umcaruje commented:
@grejppi @michaelgregorius are you sure about the 1px padding the rectrangle has? I'd find it nicer if the height of the rectrangle was the same as the height of the timebar.

grejppi commented:
@Umcaruje I'm fine with it. Not sure why but I think it would look a bit messy without the padding.

michaelgregorius commented:
Here's how it looks without any padding:
![2635-no-padding](https://cloud.githubusercontent.com/assets/9293269/15100530/8b9283b0-1574-11e6-8887-f14b26246784.png)
I have a slight preference towards the version with the padding because it looks a bit cleaner to me and makes the loop rectangle look more like an entity of its own.

In the code you can play around with different settings by changing the value of the variable `loopRectMargin` in the `TimeLineWidget::paintEvent` method. We could consider to export it as a property in the style sheet as well.

IvanMaldonado commented:
Can the padding be optional? I respect your opinion on the matter and you may be right, but I think I can fit the no padding one to look good in my themes.

Umcaruje commented:
@grejppi @michaelgregorius you're right, it does look cleaner with the padding. :+1: About making it themeable, you could make it an int qproperty.

IvanMaldonado commented:
@Umcaruje it only looks bad because the gradient on the background makes a bad contrast with the Flat aspect of the rectangle. Removing the gradients and adding borders in the timeline would make it look much better

liushuyu commented:
>it only looks bad because the gradient on the background makes a bad contrast with the Flat aspect of the rectangle. Removing the gradients and adding borders in the timeline would make it look much better

Hmm, the background of timeline widget is theme-able, so I don't  think this would become a big problem...

michaelgregorius commented:
With the latest commit the padding of the loop rectangle is now a property.

IvanMaldonado commented:
@michaelgregorius Thank you! :+1:

liushuyu commented:
@Umcaruje @IvanMaldonado  So there's no issues left, right? Let's merge this if no new occasions arise.

Umcaruje commented:
@liushuyu @michaelgregorius The code looks good to me, I'll do a test tonight, and probably merge :+1: 

One thing I'd suggest is that we actually remove the `loop_point_*` pixmaps, since they are not used anymore.

liushuyu commented:
>The code looks good to me, I'll do a test tonight, and probably merge

Okay... I have tested this and it went fine for me! :+1:

IvanMaldonado commented:
Tested and it works awesome! can the borders of the timeline be editable too? I tried adding a "border-bottom" line but it doesn't work.
Besides that I like a lot how this looks :+1:

michaelgregorius commented:
@IvanMaldonado I wonder whether it's really a bug because it also "affects" other elements (that I have not touched):
![2635-clipping](https://cloud.githubusercontent.com/assets/9293269/15193435/27b8caba-17be-11e6-9250-295a6706c5b1.png)

As you can see the beat numbers and the playhead are also rendered outside of the main area. I wonder if we should really treat it like a bug. For the loop rectangle I even like the "effect" because it gives some indication that the loop continues to the left.

However, if we decide to treat it like a bug I'd really appreciate if we created a new issue for that topic.

On a side note: the scroll bar of the automation editor and the piano roll is currently rendered on top of the time line as can be seen in the screenshot.

IvanMaldonado commented:
Alright, I may open an issue for this specific problem in the future, since other things are being worked on and this is not so important, I believe this can stay like this for now.
My last petition, can the timeline's borders be editable?

Umcaruje commented:
@michaelgregorius There is a merge conflict here because of the new theme, can you please rebase the PR?

@RebeccaDeField can you please advise which colors/style would be suitable here to fit the new theme?

RebeccaDeField commented:
> @RebeccaDeField can you please advise which colors/style would be suitable here to fit the new theme?

I believe that by removing the stroke and using either #1C4933 if you want it to be more subtle or #21A14F if you want it to be bright will work. It seems from reading over this thread that people preferred the 1px padding so I think that should be left in. I can mock it up to see what it would look like if need be.

michaelgregorius commented:
@Umcaruje I have rebased against the current master. I have redone the CSS changes for the now classic theme and have also removed the loop marker bitmaps there. The current look for the new default theme is as follows:
![2635-new_theme](https://cloud.githubusercontent.com/assets/9293269/15798764/693a394a-2a44-11e6-97aa-770d129002fe.png)
We might want to adjust the green a bit and also need to bring the gradient in the timeline back. I would very much appreciate if @RebeccaDeField could help with this because I did not see where the gradient is pulled from in the `master` branch.

@IvanMaldonado: I'd rather have this one integrated first and then we can later check for further improvements. Otherwise these pull requests become never ending stories.

RebeccaDeField commented:
@michaelgregorius The green does look a bit cloudy. My guess is that you are using a gradient that goes from white to completely transparent on top of the green? If so, the reason the color looks off is because the overlay makes the color grayer, so changing it to a gradient that is a lighter green to a dark green or removing the gradient all together might fix that.

> We also need to bring the gradient in the timeline back.

When designing this theme, I only used gradients where I felt it was absolutely necessary and I personally didn't see a reason why all of the editors such as the song editor and the automation editor did not have gradients in the grid when the timeline did. I know these are different sections of the program, but I like to keep things consistent. When @Umcaruje removed the hard coding for the gradient, some people noted that it even looked nice with the old theme (but didn't match as that theme uses gradients on everything) and I noticed that other people are using the timeline without gradients in their themes as well. I also did some research and noted that many other DAWs don't use gradients in the timeline.

I  would prefer to leave the gradient out in that section, but, this is Open Source and I don't mind if people make changes that I'm not entirely in agreement with because this is the default theme. I would also like to hear what some other people in the LMMS team think? :)

Thanks for working so hard on this!

liushuyu commented:
IMO, the flat timeline fits better with @RebeccaDeField 's theme. I think maybe it would be good to add yet another exposure to CSS that can control whether gradients are enabled.

Umcaruje commented:
I also agree that the timeline would look a lot better when flat. Good job on the rebasing @michaelgregorius :+1:

michaelgregorius commented:
@RebeccaDeField No need to thank me. Most of the hard work for the new theme was done on your side anyway.

I have adjusted the CSS to use no more gradients and borders as proposed by @RebeccaDeField. I have removed the borders and have used the less subtle `#21A14F` for the active state and `#41464A` for the inactive state. Here's how it look with the latest commit:
![2635-new_theme_active_inactive](https://cloud.githubusercontent.com/assets/9293269/15804893/105119cc-2b19-11e6-86f6-b7ae7cb4d37b.png)

I propose to merge this pull request and to do some potential remaining changes in other commits/pull requests.

@liushuyu The classic theme uses a gradient for the time line so it should already be possible to use gradients via CSS.

RebeccaDeField commented:
@michaelgregorius Before we merge, can we adjust the inactive color to #3B424A? The current color has a slightly different hue and saturation than the gray on the bar that is is sitting on, so it clashes a bit. A good way to get the right shade is to color-pick the main color in the windows and just change the value (or lightness/darkness.)

This can always be done after the merge if need be.

michaelgregorius commented:
@RebeccaDeField I have added a new commit that changes the color of inactive loops as proposed.

RebeccaDeField commented:
@michaelgregorius Thanks, I think we can make any additional changes later as you said.

michaelgregorius commented:
I will merge this one in the next few days if there are no objections.

Umcaruje commented:
@michaelgregorius sounds fine to me :+1: Just remember to squash and merge on github (When merging the PR, you have a dropdown menu next to the merge button where you can select that).

RebeccaDeField commented:
:+1:
